apps/emqx_durable_storage/src/emqx_ds_storage_layer.erl

```diff
@@ -297,13 +301,14 @@ store_batch(Shard, Messages, Options) ->
     [{emqx_ds:time(), emqx_types:message()}],
     emqx_ds:message_store_opts()
 ) -> {ok, cooked_batch()} | ignore | emqx_ds:error(_).
-prepare_batch(Shard, Messages = [{Time, _Msg} | _], Options) ->
+prepare_batch(Shard, Messages = [_ | _], Options) ->
     %% NOTE
     %% We assume that batches do not span generations. Callers should enforce this.
     ?tp(emqx_ds_storage_layer_prepare_batch, #{
         shard => Shard, messages => Messages, options => Options
     }),
-    {GenId, #{module := Mod, data := GenData}} = generation_at(Shard, Time),
+    GenId = generation_current(Shard),
```

Comment:
I suspect this could introduce message duplication under _snapshot transfer + log replay_ scenarios. I.e. imagine a following sequence of entries in the logs:
1. `{T1 = 42, Msg1}`
2. `{T2 = 43, Msg2}`
3. `{T3 = 44, add_generation}`
4. `{T4 = 45, Msg3}`

If they are reapplied on top of the storage state that _already had them applied_, couldn't `Msg1` and `Msg2` end up both in previous and in the current generation?

Reply:
Yes, it will happen. But running prepare in the main state machine code should be removed. Prepare should run on the leader (in aux callback), and add a cooked batch to the raft log. Committing cooked batch to the storage is idempotent.